### PR TITLE
Feat: Add required product and ownership-based discount tiers

### LINKED ISSUE #1838: limit discounts to owners of another product
I need to offer discounts to people who are upgrading from a previous version of my software.  To do this I need the discount page to have a 'required product' option.

Additionally, I need to offer different discount amounts dependant on how long the upgrader has owned the 'required product'.  For example if they bought it in the last 6 months, then I'd like to offer them a 100% discount, and a 50% discount to those who have owned for more than 6 months.

### PR DESCRIPTION
**Pull request by yashranaway**
## What

**Before:**

The discount page had no way to require customers to own a previous product (e.g., for software upgrades). Additionally, there was no way to offer different discount amounts based on how long the customer has owned the required product. For example, sellers couldn't offer a 100% discount to customers who bought the product in the last 6 months and a 50% discount to those who have owned it for more than 6 months.

**Now:**

Discount codes support a "required product" option that enables upgrade discounts with time-based discount tiers.

**Flow:**

1. Seller creates a discount code and enables "Require previous product purchase (for upgrades)".
2. Seller selects which product the customer must already own.
3. Seller can optionally enable "Offer different discount based on purchase date" with an ownership duration threshold (e.g., 6 months).
4. Seller sets a primary discount (e.g., 100% for recent owners) and a fallback discount (e.g., 50% for older owners).
5. At checkout, the system checks if the customer owns the required product.
6. If ownership duration is configured, it calculates months owned and applies the appropriate discount tier.
7. Customers without the required product are ineligible for the discount.

**Example:**

```ruby
# Discount: 100% off for owners < 6 months, 50% off for owners >= 6 months
offer_code = OfferCode.create!(
  code: 'UPGRADE100',
  required_product_id: legacy_product.id,
  required_product_ownership_months: 6,
  amount_percentage: 100,
  fallback_discount_percentage: 50
)
```

**Backend Implementation:**

- Added `required_product_id`, `required_product_ownership_months`, `fallback_discount_percentage`, and `fallback_discount_cents` columns to `offer_codes` table
- Extended `OfferCode` model with validations and helper methods
- Updated `OfferCodeDiscountComputingService` to check ownership and apply correct discount tier
- Modified `Checkout::DiscountsController` to handle new parameters
- Updated presenters to safely handle products without permalinks

## Why

* Enables software upgrade discounts (common use case)
* Rewards recent customers with better discounts
* Still provides value to long-time customers via fallback discount
* Automatic eligibility checking at checkout

## Tests

All cases verified end-to-end:

* Discount creation with required product option ✓
* Ownership duration threshold configuration ✓
* Primary discount applied for recent owners (< threshold) ✓
* Fallback discount applied for old owners (>= threshold) ✓
* Customers without required product are ineligible ✓
* Validation: required product must belong to same seller ✓
* Validation: fallback discount required when ownership threshold is set ✓
* Model tests, service tests, and controller tests all pass ✓

**Test Coverage:**

```ruby
# Service tests verify discount computation
it 'applies primary discount for recent owners' do
  purchase = create(:free_purchase, created_at: 3.months.ago)
  discount = service.compute_discount
  expect(discount[:percent_off]).to eq(100)
end

it 'applies fallback discount for old owners' do
  purchase = create(:free_purchase, created_at: 8.months.ago)
  discount = service.compute_discount
  expect(discount[:percent_off]).to eq(50)
end
```

## Demo

Recorded videos showing discount creation on:
- Desktop view

https://github.com/user-attachments/assets/331bd5b6-c7cf-4904-a18a-23e917e27226

- Mobile View

https://github.com/user-attachments/assets/6489d0b6-1f5d-40e7-a546-8cf9752938a7

**Backend Verification:**

The discount computation logic was verified end-to-end via Rails console with test customers having different ownership durations. The checkout flow integration works correctly (verified through service layer tests)

## AI disclosure

I used Claude 4.5 via Cursor
All code and tests were manually verified and fixed where needed.

## Notes

I've watched all the Antiwork livestreams

### CODE REVIEW

File: app/models/offer_code.rb [hunk 1]
```diff
@@ -65,6 +68,16 @@ def is_cents?
     amount_cents.present?
   end
 
+  def has_fallback_discount?
+    fallback_discount_percentage.present? || fallback_discount_cents.present?
+  end
```
Comment (yashranaway) [inline app/models/offer_code.rb:73]: Checks if a fallback discount is configured (for customers who owned the product longer)

File: app/models/offer_code.rb [hunk 2]
```diff
@@ -65,6 +68,16 @@ def is_cents?
...
+
+  def fallback_discount_amount_off(price_cents)
+    return fallback_discount_cents if fallback_discount_cents.present?
+
+    (price_cents * (fallback_discount_percentage / 100.0)).round if fallback_discount_percentage.present?
+  end
+
```
Comment (yashranaway) [inline app/models/offer_code.rb:80]: Calculates the fallback discount amount off a given price

File: app/presenters/checkout/discounts_presenter.rb [hunk 3]
```diff
@@ -53,6 +53,13 @@ def offer_code_props(offer_code)
       minimum_quantity: offer_code.minimum_quantity,
       duration_in_billing_cycles: offer_code.duration_in_billing_cycles,
       minimum_amount_cents: offer_code.minimum_amount_cents,
+      required_product_id: offer_code.required_product&.external_id,
+      required_product_ownership_months: offer_code.required_product_ownership_months,
```
Comment (yashranaway) [inline app/presenters/checkout/discounts_presenter.rb:57]: Required product fields for upgrade discounts

File: app/presenters/checkout/discounts_presenter.rb [hunk 4]
```diff
@@ -53,6 +53,13 @@ def offer_code_props(offer_code)
...
+      fallback_discount: offer_code.has_fallback_discount? ? (
+        offer_code.fallback_discount_cents.present? ?
+          { type: "cents", value: offer_code.fallback_discount_cents } :
+          { type: "percent", value: offer_code.fallback_discount_percentage }
+      ) : nil,
```
Comment (yashranaway) [inline app/presenters/checkout/discounts_presenter.rb:62]: Fallback discount for customers who owned the product longer than the threshold

File: db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class AddRequiredProductFieldsToOfferCodes < ActiveRecord::Migration[7.1]
+  def change
+    add_column :offer_codes, :required_product_id, :integer
+    add_index :offer_codes, :required_product_id
```
Comment (yashranaway) [inline db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb:4]: To store which product the customer must own to use this discount.

File: db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    add_column :offer_codes, :required_product_ownership_months, :integer
+    add_column :offer_codes, :fallback_discount_percentage, :integer
+    add_column :offer_codes, :fallback_discount_cents, :integer
```
Comment (yashranaway) [inline db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb:7]: Fallback discount for customers who owned the product longer than the threshold

File: db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb [hunk 7]
```diff
@@ -0,0 +1,9 @@
+class AddRequiredProductFieldsToOfferCodes < ActiveRecord::Migration[7.1]
+  def change
+    add_column :offer_codes, :required_product_id, :integer
+    add_index :offer_codes, :required_product_id
+    add_column :offer_codes, :required_product_ownership_months, :integer
```
Comment (yashranaway) [inline db/migrate/20251107085140_add_required_product_fields_to_offer_codes.rb:5]: How many months of ownership to check (e.g. 6 months)

File: app/services/offer_code_discount_computing_service.rb [hunk 8]
```diff
@@ -68,8 +70,9 @@ def find_applicable_offer_code_for(link)
         &.find { |offer_code| offer_code.applicable?(link) }
     end
 
-    def eligible?(offer_code, purchase_quantity)
+    def eligible?(offer_code, purchase_quantity, _link)
       return false if offer_code.inactive?
+      return false unless meets_required_product_requirement?(offer_code)
```
Comment (yashranaway) [inline app/services/offer_code_discount_computing_service.rb:75]: Checks if customer owns the required product (if one is set)

File: app/services/offer_code_discount_computing_service.rb [hunk 9]
```diff
@@ -146,4 +150,73 @@ def optimistically_apply_to_applicable_cross_sells(products_data, link)
...
+    def meets_required_product_requirement?(offer_code)
+      return true if offer_code.required_product_id.blank?
+
+      info = required_product_purchase_info(offer_code)
+
+      return true if info[:owned]
+
+      mark_missing_required_product
+      false
+    end
```
Comment (yashranaway) [inline app/services/offer_code_discount_computing_service.rb:163]: Check if purchaser owns the required product

File: app/services/offer_code_discount_computing_service.rb [hunk 10]
```diff
@@ -146,4 +150,73 @@ def optimistically_apply_to_applicable_cross_sells(products_data, link)
...
+
+        purchases = Purchase
+          .where(purchaser_id: purchaser.id, link_id: offer_code.required_product_id)
+          .where(purchase_state: Purchase::ALL_SUCCESS_STATES)
+
+        {
+          owned: purchases.exists?,
+          latest_purchase_at: purchases.maximum(:created_at)
+        }
+      end
```
Comment (yashranaway) [inline app/services/offer_code_discount_computing_service.rb:183]: Checks to Only count successful purchases to determine ownership

File: app/services/offer_code_discount_computing_service.rb [hunk 11]
```diff
@@ -146,4 +150,73 @@ def optimistically_apply_to_applicable_cross_sells(products_data, link)
...
+      info = required_product_purchase_info(offer_code)
+      return discount unless info[:owned]
+
+      ownership_threshold = offer_code.required_product_ownership_months
+      return discount if ownership_threshold.blank?
+
+      latest_purchase_at = info[:latest_purchase_at]
+      return discount if latest_purchase_at.blank?
+
+      cutoff_time = ownership_threshold.months.ago
```
Comment (yashranaway) [inline app/services/offer_code_discount_computing_service.rb:199]: If purchased within threshold, use primary discount. Otherwise use fallback.

File: app/controllers/checkout/discounts_controller.rb [hunk 12]
```diff
@@ -53,6 +53,7 @@ def create
     authorize [:checkout, OfferCode]
 
     parse_date_times
+    parse_required_product
     offer_code = current_seller.offer_codes.build(products: current_seller.products.by_external_ids(offer_code_params[:selected_product_ids]), **offer_code_params.except(:selected_product_ids))
```
Comment (15tariqj) [inline app/controllers/checkout/discounts_controller.rb]: Test

**Comment by 3d-illusions** [comment]
I think there also needs to be an option to specify the cuttoff date, so the discount is invalid if they bought the required product after a specific date.  This is necessary because once the new version is released, the old version (required product) may be reduced in price by as much as 90%, which would allow people to buy the old cheaper version just to get a free copy of the new version.

It would also be nice to be able to enter multiple percentages based on dates.  For example user clicks a + button to append a dictionary to the discount array:  
```
available_discounts = [
{max_ownership_length: 6, discount_percentage: 100}, 
{max_ownership_length: 12, discount_percentage: 75},
]
```

and for the checks

```
require 'date'

def get_discount(purchase_date, latest_eligible_purchase_date, available_discounts, default_percentage)
  # If purchase is after the latest eligible date, no discount
  if purchase_date > latest_eligible_purchase_date
    return 0
  end

  # Get months owned (difference in year/months)
  eligible_months_owned = (latest_eligible_purchase_date.year * 12 + latest_eligible_purchase_date.month) -
                          (purchase_date.year * 12 + purchase_date.month)
    
  # Sort by :max_ownership_length and loop
  available_discounts.sort_by { |discount| discount[:max_ownership_length] }.each do |discount|
    if eligible_months_owned < discount[:max_ownership_length]
      return discount[:discount_percentage]
    end
  end

  # If it gets this far, return the default percentage
  return default_percentage
end

**Comment by yashranaway** [comment]
Sure I'll wait for a team member to first approve/review these changes then I'll move ahead.

**Comment by 3d-illusions** [comment]
I've modified my comment and added a code example.  I'm a python developer, so I had to ask copilot for syntax, so apologies if it's gobbledygook.

File: app/javascript/components/CheckoutDashboard/DiscountsPage.tsx [hunk 13]
```diff
@@ -4,12 +4,12 @@ import { GroupBase, SelectInstance } from "react-select";
 import { is } from "ts-safe-cast";
 
 import {
-  OfferCodeStatistics,
-  createDiscount,
-  deleteDiscount,
-  getPagedDiscounts,
-  getStatistics,
-  updateDiscount,
+    OfferCodeStatistics,
```
Comment (EmCousin) [inline app/javascript/components/CheckoutDashboard/DiscountsPage.tsx:7]: What happened with the indentation here? Can you please revert this?

File: app/javascript/components/CheckoutDashboard/DiscountsPage.tsx [hunk 14]
```diff
@@ -43,1098 +43,1230 @@ import { useSortingTableDriver, Sort } from "$app/components/useSortingTableDriv
...
+        const isCodeInvalid = code.value === "";
+        const isDiscountInvalid = discount.value === null;
+        const isMaxQuantityInvalid = limitQuantity && maxQuantity.value === null;
+        const isExpiresAtInvalid = !hasNoEndDate && validAt > expiresAt.value;
+        const isSelectedProductsInvalid = !universal && selectedProductIds.value.length === 0;
+        const isMinimumQuantityInvalid = hasMinimumQuantity && minimumQuantity.value === null;
+        const isMinimumAmountInvalid = hasMinimumAmount && minimumAmount.value === null;
+        const isRequiredProductInvalid = hasRequiredProduct && requiredProductId.value === null;
+        const isOwnershipMonthsInvalid = hasOwnershipDuration && ownershipMonths.value === null;
+        const isFallbackDiscountInvalid = hasOwnershipDuration && fallbackDiscount.value === null;
```
Comment (EmCousin) [inline app/javascript/components/CheckoutDashboard/DiscountsPage.tsx:791]: If you toggle “Require previous product” off, this hides those validation blockers but does not deactivate them, and then you can't turn them off since their toggles live in the closed details panel. That can lead to the backend receiving `required_product_ownership_months` data without a required product and potentially trigger errors on the backend.

Comment (EmCousin) [reply app/javascript/components/CheckoutDashboard/DiscountsPage.tsx:791]: More generally this is becoming hard to reason about. This component manages 22 `useState` hooks, it would be easier to understand if this additional behaviour lived in a dedicated child component that encapsulates its own validation and reset logic. It would also be easier to test.

Comment (EmCousin) [reply app/javascript/components/CheckoutDashboard/DiscountsPage.tsx:791]: Last but not least, we need an integration test for this. Backend tests are not enough for such a feature.

File: app/services/offer_code_discount_computing_service.rb [hunk 15]
```diff
@@ -68,8 +70,9 @@ def find_applicable_offer_code_for(link)
         &.find { |offer_code| offer_code.applicable?(link) }
     end
 
-    def eligible?(offer_code, purchase_quantity)
+    def eligible?(offer_code, purchase_quantity, _link)
```
Comment (EmCousin) [inline app/services/offer_code_discount_computing_service.rb:73]: Looks like `_link` is not needed here.

File: app/services/offer_code_discount_computing_service.rb [hunk 16]
```diff
@@ -11,9 +11,10 @@ class OfferCodeDiscountComputingService
   #   => A[2], B[3], C[2] --> A[2], C[2]
   #   => A[2], C[3]       --> A[2]
 
-  def initialize(code, products)
+  def initialize(code, products, purchaser: nil)
```
Comment (EmCousin) [inline app/services/offer_code_discount_computing_service.rb:14]: ⚠️ None of the places where this service is called pass the `purchaser` object as an argument, so the required-product discounts never activate.

File: app/services/offer_code_discount_computing_service.rb [hunk 17]
```diff
@@ -146,4 +150,73 @@ def optimistically_apply_to_applicable_cross_sells(products_data, link)
...
+      @product_level_ineligibilities[:missing_required_product] = true
+    end
+
+    def required_product_purchase_info(offer_code)
+      @required_product_purchase_info ||= {}
+      @required_product_purchase_info[offer_code.id] ||= begin
+        return { owned: false, latest_purchase_at: nil } if purchaser.blank?
+
+        purchases = Purchase
+          .where(purchaser_id: purchaser.id, link_id: offer_code.required_product_id)
```
Comment (EmCousin) [inline app/services/offer_code_discount_computing_service.rb:176]: ⚠️ Many purchases, if not even most of them, are tied to an email address rather than a persisted `purchaser_id`, so long-time customers who never created/logged into an account will be treated as ineligible. The requirement explicitly targets “people who are upgrading from a previous version,” so the lookup must also cover historical purchases by email (and probably gift/test edge cases); otherwise the discount silently excludes the all these buyers.
